### PR TITLE
Optimize column vector filter

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #6839

Problem Summary:

### What is changed and how it works?

rewrite `ColumnVector<T>::filter` with `avx2`

perf test as follows (larger is better):

```shell
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 11:55:31]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 1 30
Test int64-filter rows=10000 columns=1 seconds=30
FilterV1: 117192    
FilterV2: 157265    
FilterV3: 157951    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 11:56:09]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 10 30
Test int64-filter rows=10000 columns=10 seconds=30
FilterV1: 157057    
FilterV2: 290671    
FilterV3: 95683     
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 11:56:54]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 50 30
Test int64-filter rows=10000 columns=50 seconds=30
FilterV1: 408640    
FilterV2: 803905    
FilterV3: 398920    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 11:57:32]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 100 30
Test int64-filter rows=10000 columns=100 seconds=30
FilterV1: 532940    
FilterV2: 983276    
FilterV3: 582252    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 11:58:07]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 101 30
Test int64-filter rows=10000 columns=101 seconds=30
FilterV1: 833012    
FilterV2: 1339852   
FilterV3: 1159898
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Columns/ColumnVector.cpp [hunk 1]
```diff
@@ -203,6 +203,69 @@ void ColumnVector<T>::insertRangeFrom(const IColumn & src, size_t start, size_t
...
+inline void doFilterAligned(const UInt8 *& filt_pos, const UInt8 *& filt_end_aligned, const T *& data_pos, Container & res_data)
+{
+    while (filt_pos < filt_end_aligned)
+    {
+        UInt64 mask = ToBits64(filt_pos);
+        const UInt8 prefix_to_copy = prefixToCopy(mask);
+
+        if (0xFF != prefix_to_copy)
+        {
+            res_data.insert(data_pos, data_pos + prefix_to_copy);
```
Comment (solotzg) [inline dbms/src/Columns/ColumnVector.cpp:244]: Is it possible to find every pair of start / end position of continuous section to copy?

Comment (Lloyd-Pottiger) [reply dbms/src/Columns/ColumnVector.cpp:244]: Seems do not have such function to achieve this.

Comment (Lloyd-Pottiger) [reply dbms/src/Columns/ColumnVector.cpp:244]:
``` cpp
// 1
UInt64 mask = ToBits64(filt_pos);
const UInt8 prefix_to_copy = prefixToCopy(mask);
if (0xFF != prefix_to_copy)
{
    res_data.insert(data_pos, data_pos + prefix_to_copy);
}
else
{
    const uint8_t suffix_to_copy = suffixToCopy(mask);
    if (0xFF != suffix_to_copy)
    {
        res_data.insert(data_pos + SIMD_BYTES - suffix_to_copy, data_pos + SIMD_BYTES);
    }
    else
    {
        while (mask)
        {
            size_t index = __builtin_ctzll(mask);
            res_data.push_back(data_pos[index]);
            mask &= mask - 1;
        }
    }
}

// 2
UInt64 mask = ToBits64(filt_pos);
if (0xFFFFFFFFFFFFFFFFULL == mask)
{
    res_data.insert(data_pos, data_pos + SIMD_BYTES);
}
else
{
    int processed_bytes = 0;
    while (mask)
    {
        const int prefix_to_copy = __builtin_ctzll(~mask);
        if (prefix_to_copy)
        {
            res_data.insert(data_pos + processed_bytes, data_pos + processed_bytes + prefix_to_copy);
            processed_bytes += prefix_to_copy;
            mask >>= prefix_to_copy;
        }
        else
        {
            const int suffix_to_pass = __builtin_ctzll(mask);
            processed_bytes += suffix_to_pass;
            mask >>= suffix_to_pass;
        }
    }
}
```

```shell 
# 1
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:25:44]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 1 30
Test int64-filter rows=10000 columns=1 seconds=30
Filter/s: 156726    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:26:25]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 10 30
Test int64-filter rows=10000 columns=10 seconds=30
Filter/s: 336723    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:27:06]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 50 30
Test int64-filter rows=10000 columns=50 seconds=30
Filter/s: 632002    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:27:45]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 100 30
Test int64-filter rows=10000 columns=100 seconds=30
Filter/s: 624887    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:28:29]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 101 30
Test int64-filter rows=10000 columns=101 seconds=30
Filter/s: 526091    

# 2
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:17:21]$ ninja column_vector_perftest
[0/2] Re-checking globbed directories...
[3/3] Linking CXX executable dbms/src/Columns/tests/column_vector_perftest
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:20:26]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 1 30
Test int64-filter rows=10000 columns=1 seconds=30
Filter/s: 151959    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:20:58]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 10 30
Test int64-filter rows=10000 columns=10 seconds=30
Filter/s: 95072     
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:22:07]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 50 30
Test int64-filter rows=10000 columns=50 seconds=30
Filter/s: 328683    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:22:41]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 100 30
Test int64-filter rows=10000 columns=100 seconds=30
Filter/s: 474129    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:23:21]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 101 30
Test int64-filter rows=10000 columns=101 seconds=30
Filter/s: 1030307   
```

File: dbms/src/Columns/tests/column_vector_perftest.cpp [hunk 2]
```diff
@@ -88,6 +100,38 @@ void testScatter(int num_rows, int num_columns, int seconds)
...
+
+    auto filter_func = [&] {
+        while (!stop_flag.load(std::memory_order_relaxed))
+        {
+            src->filter(filter, -1);
+            counter.fetch_add(1, std::memory_order_relaxed);
+        }
+    };
+
+    std::thread t(filter_func);
```
Comment (solotzg) [inline dbms/src/Columns/tests/column_vector_perftest.cpp:122]: Maybe we just need one thread and test the time costs of handling same filter / data by different method.

File: dbms/src/Columns/tests/column_vector_perftest.cpp [hunk 3]
```diff
@@ -88,6 +100,38 @@ void testScatter(int num_rows, int num_columns, int seconds)
         << std::endl;
 }
 
+template <typename T>
+void testFilter(int num_rows, int num_columns, int seconds)
+{
+    ColumnPtr src = buildColumn<T>(num_rows);
+    auto filter = buildFilter(num_rows, num_columns);
```
Comment (solotzg) [inline dbms/src/Columns/tests/column_vector_perftest.cpp:107]: Can you guarantee that the generated filter is same for different test cases?

File: dbms/src/Columns/ColumnVector.cpp [hunk 4]
```diff
@@ -203,6 +203,70 @@ void ColumnVector<T>::insertRangeFrom(const IColumn & src, size_t start, size_t
...
+
+inline UInt8 suffixToCopy(UInt64 mask)
+{
+    const auto prefix_to_copy = prefixToCopy(~mask);
+    return prefix_to_copy >= 64 ? prefix_to_copy : 64 - prefix_to_copy;
+}
+
+} // namespace
+
+template <typename T, typename Container, size_t SIMD_BYTES>
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Columns/ColumnVector.cpp]: change to `template <typename T, typename Container, size_t SIMD_BYTES = 64>`

File: dbms/src/Columns/ColumnVector.cpp [hunk 5]
```diff
@@ -203,6 +203,70 @@ void ColumnVector<T>::insertRangeFrom(const IColumn & src, size_t start, size_t
...
+        return 0;
+    if (mask == all_match)
+        return 64;
+    /// Row with index 0 correspond to the least significant bit.
+    /// So the length of the prefix to copy is 64 - #(leading zeroes).
+    const UInt64 leading_zeroes = __builtin_clzll(mask);
+    if (mask == ((all_match << leading_zeroes) >> leading_zeroes))
+        return 64 - leading_zeroes;
+    else
+        return 0xFF;
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Columns/ColumnVector.cpp:223]: No, for `64 - leading_zeroes` can never be `0xFF`, I think it is enough.

File: dbms/src/Columns/tests/column_vector_perftest.cpp (hunk 2 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Columns/tests/column_vector_perftest.cpp]: Have update the perf test and result in pr description, PTAL.

**Pull request closed without merging**